packages/flutter/lib/src/rendering/selection.dart

```diff
@@ -0,0 +1,607 @@
+  previous,
+  /// Selection ends in this [Selectable].
+  ///
+  /// Part of the [Selectable] may or may not be selected, but there is still
+  /// content to select forward or backward.
+  ///
+  /// {@macro flutter.rendering.selection.SelectionResult.footNote}
+  end,
+  /// The result can't be determined in this frame.
+  ///
+  /// This is typically used when the subtree is scrolling to reveal more
+  /// content.
+  ///
+  /// {@macro flutter.rendering.selection.SelectionResult.footNote}
+  // See `_SelectableRegionState._triggerSelectionEndEdgeUpdate` for how this
+  // result affects the selection.
+  pending,
+  /// There is no result for the selection event.
+  ///
+  /// This is used when a selection result is not applicable, e.g.
+  /// [SelectAllSelectionEvent], [ClearSelectionEvent], and
+  /// [SelectWordSelectionEvent].
+  none,
+}
+
+/// The abstract interface to handle [SelectionEvent]s.
+///
+/// This interface is extended by [Selectable] and [SelectionContainerDelegate]
+/// and is typically not use directly.
+///
+/// {@template flutter.rendering.SelectionHandler}
+/// This class returns a [SelectionGeometry] as its [value], and is responsible
+/// to notify its listener when its selection geometry has changed as the result
+/// of receiving selection events.
+/// {@endtemplate}
+abstract class SelectionHandler implements ValueListenable<SelectionGeometry> {
+  /// Marks this handler to be responsible for pushing leader layers for the
+  /// selection handles.
+  ///
+  /// This handler is responsible for pushing the leader layers with the
+  /// given layer links if they are not null. It is possible that only one layer
+  /// is non-null if this handler is only responsible for pushing one layer
+  /// link.
+  ///
+  /// The `startHandle` needs to be placed at the visual location of selection
+  /// start, the `endHandle` needs to be placed at the visual location of selection
+  /// end. Typically, the visual locations should be the same as
+  /// [SelectionGeometry.startSelectionPoint] and
+  /// [SelectionGeometry.endSelectionPoint].
+  void pushHandleLayers(LayerLink? startHandle, LayerLink? endHandle);
+
+  /// Gets the selected content in this object.
+  ///
+  /// Return `null` if nothing is selected.
+  SelectedContent? getSelectedContent();
+
+  /// Handles the [SelectionEvent] sent to this object.
+  ///
+  /// The subclasses need to update their selections or delegate the
+  /// [SelectionEvent]s to their subtrees.
+  ///
+  /// The `event`s are subclasses of [SelectionEvent]. Check
+  /// [SelectionEvent.type] to determine what kinds of event are dispatched to
+  /// this handler and handle them accordingly.
+  ///
+  /// See also:
+  ///  * [SelectionEventType], which contains all of the possible types.
+  SelectionResult dispatchSelectionEvent(SelectionEvent event);
+}
+
+/// The selected content in a [Selectable] or [SelectionHandler].
+// TODO(chunhtai): Add more support for rich content.
```

Comment:
Nit: Consider creating and linking issues for this and other TODOs in this PR, if you don't already have a good way to keep track of them.